nibabel/brainvoyager/bv.py

```diff
@@ -0,0 +1,952 @@
+    fields (e.g. the ``prts`` field of the VTC header).
+
+    Parameters
+    ----------
+    hdr_dict_proto: tuple
+        tuple of format described in Notes of :func:`parse_BV_header`
+    hdr_dict: OrderedDict
+       hdr_dict that contains the fields and values to for the respective
+       BV file format.
+    parent_hdr_dict: OrderedDict
+       When parse_BV_header() is called recursively the already filled
+       (parent) hdr_dict is passed to give access to n_fields_name fields
+       outside the current scope (see below).
+
+    Returns
+    -------
+    binaryblock : bytes
+        Binary representation of header ready for writing to file.
+    """
+    binary_parts = []
+    for name, format, def_or_name in hdr_dict_proto:
+        value = hdr_dict[name]
+        # handle zero-terminated strings
+        if format == 'z':
+            part = value + b'\x00'
+        # handle array fields
+        elif isinstance(format, tuple):
+            # check the length of the array to expect
+            if def_or_name in hdr_dict:
+                n_values = hdr_dict[def_or_name]
+            else:
+                n_values = parent_hdr_dict[def_or_name]
+            sub_parts = []
+            for i in range(n_values):
+                sub_parts.append(pack_BV_header(format, value[i], hdr_dict))
+            part = b''.join(sub_parts)
+        # handle conditional fields
+        elif isinstance(def_or_name, tuple):
+            if hdr_dict[def_or_name[1]] == def_or_name[2]:
+                part = pack('<' + format, value)
+            else:
+                continue
+        else:
+            part = pack('<' + format, value)
+        binary_parts.append(part)
+    return b''.join(binary_parts)
+
+
+def calc_BV_header_size(hdr_dict_proto, hdr_dict, parent_hdr_dict=None):
+    """Calculate the binary size of a hdr_dict for a BV file format header.
+
+    This function can be (and is) called recursively to iterate through nested
+    fields (e.g. the prts field of the VTC header).
+
+    Parameters
+    ----------
+    hdr_dict_proto: tuple
+        tuple of format described in Notes of :func:`parse_BV_header`
+    hdr_dict: OrderedDict
+       hdr_dict that contains the fields and values to for the respective
+       BV file format.
+    parent_hdr_dict: OrderedDict
+       When parse_BV_header() is called recursively the already filled
+       (parent) hdr_dict is passed to give access to n_fields_name fields
+       outside the current scope (see below).
+
+    Returns
+    -------
+    hdr_size : int
+        Size of header when packed into bytes ready for writing to file.
+    """
+    hdr_size = 0
+    for name, format, def_or_name in hdr_dict_proto:
+        value = hdr_dict[name]
+        # handle zero-terminated strings
+        if format == 'z':
+            hdr_size += len(value) + 1
+        # handle array fields
+        elif isinstance(format, tuple):
```

Comment:
I agree that this is not the most beautiful style here. However, there are a lot of weird edge cases with BV headers and I fear that while trying to make things more general one would make things really hard to understand, too. At least this repetition is confined to these module level functions in bv.py and do not repeat for different file formats.
I will give it some more thought though.